### PR TITLE
New Tcp Forwarder Output plugin.

### PR DESCRIPTION
**Pull request by tuier**
This can be use to forward metric to a centralized endpoint (telegraf or not)

Sample use case:

```
Considering a Pool of server with telegraf and tcp_forwarder
enabled, who will send data to some other sever with tcp_listener.

This allow to have a better security, as credential for "real"
output is not in every server of the Pool, and will allow to rotate
credential much easier.
```

### CODE REVIEW

File: plugins/outputs/tcp_forwarder/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+# Tcp Forwarder Output Plugin
+
+This plugin will send all metrics through TCP in the chosen format, this can be 
+use by example with tcp listener input plugin
+
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/README.md]: please put a configuration example here

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+var sampleConfig = `
+  ## TCP server/endpoint to send metrics to.
+  servers = ["localhost:8089"]
+  ## timeout in seconds for the write connection
+  timeout = 2
+  ## reconnect before every push
+  reconnect = false
+  ## Data format to consume.
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: should be Data format to _output_.

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	"github.com/influxdata/telegraf/plugins/outputs"
+	"github.com/influxdata/telegraf/plugins/serializers"
+)
+
+// TCPForwarder structure for configuration and server
+type TCPForwarder struct {
+	Servers    []string
+	Timeout    int
+	DataFormat string
+	Reconnect  bool
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: you never use the `Reconnect` bool

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+		bp = append(bp, sMetrics...)
+	}
+	// TODO should we add a join function in serialiser ?
+	points := strings.Join(bp, "\n") + "\n"
+
+	// This will get set to nil if a successful write occurs
+	err := errors.New("Could not write to any tcp endpoint\n")
+
+	// Send data to a random endpoint
+	p := rand.Perm(len(t.Servers))
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: you have not documented that this output selects a random server to write to, thereby load-balancing the output servers defined.

Was this done on purpose? I'm not sure it should be, since it's a generic TCP output. Where we have done this in the past is only for output sinks that are clustered, and thus only need one IP address to be written to.

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	for _, n := range p {
+		endpoint := t.Servers[n]
+		if value, ok := t.conns[endpoint]; ok && value != nil {
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: you should be setting a WriteDeadline on the tcp connection before trying to write to it so that it can't hang

**Comment by sparrc** [comment]
you should put a sync.Mutex on the plugin struct, lock it while gathering and closing so that they can't be called at the same time.

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go (hunk 4 above)
Comment (tuier) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: It's made in purpose.
This follow the way the Graphite UDP output use.
I agree that TCP may not need it, but this give the options to people to set multiple endpoints if they wanted.

@sparrc :  Do you want me to remove it ?

Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: yes, please remove

**Comment by tuier** [comment]
@sparrc any more work needed on this PR ?

**Comment by sparrc** [comment]
I still need to do a final review and test it. There is currently a large backlog of PRs and this may take a while, so please be patient.

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+	"sync"
+	"time"
+)
+
+// TCPForwarder structure for configuration and server
+type TCPForwarder struct {
+	sync.Mutex
+
+	Server     string
+	Timeout    string
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: change the type of `Timeout` to `internal.Duration`, then you don't need to parse it, the toml parsing will deal with verifying that it's a valid timestamp.

in the config file it will still look like `timeout = "5s"`

File: plugins/outputs/tcp_forwarder/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+```toml
+[[outputs.instrumental]]
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/README.md]: outputs.instrumental ?

**Comment by theist** [comment]
Hi I'm testing this plugin searching for ways to "proxy" telegraf in firewalled servers (I'm testing it along other solutions like use of haproxy and the http service input plugin done in #1407) 

I'm using this config:
```
[[outputs.tcp_forwarder]]
  server = "localhost:8186"
  timeout = "5s"
  reconnect = false
  data_format = "influx"
```

I placed recconect to false because it feels a bit of overhead according the help text `  ## force reconnection before every push`

When I setup a pair of tcp_input / tcp_output two things happen

First of all I cannot start the "tcp_output_ed" if I don't start the "tcp_input_ed" terrafor ends without gather any metrics. Perhaps will we great that telegraf start collecting data and add it to the buffer even if there's no tcp endpoint listening on the other side. But I can workarround that and it is not my main concern

My problem is that if I drop the "telegraf proxy" the proxied with the tcp_output starts to accumulate data in buffer as expected, but when the proxy with the tcp_input restarts again it seems that the conection is never retried and the proxied keeps logging

```
ERROR: write tcp 127.0.0.1:59778->127.0.0.1:8186: use of closed network connection
2016/09/04 12:53:40 Error writing to output [tcp_forwarder]: Could not write to tcp endpoint
```

Is the connection ever retried? Sorry I tried to look at the code but golang seems Greek to me

**Comment by tuier** [comment]
For the first issues: when starting the plugin a connect function is called to make sure it can connect this is a default behavior of telegraf

@sparrc Should we change that to the suggestion ? gathering metric before we can reach a endpoint, by not passing the first connection error ?

@theist Can you share the whole configuration file for both side ?

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+func (t *TCPForwarder) Write(metrics []telegraf.Metric) error {
+	// reconnect if needed
+	if err := t.reconnect(); err != nil {
+		return err
+	}
+	// Prepare data
+	t.Lock()
+	defer t.Unlock()
+
+	var bp []string
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: rather than making a slice of strings and then joining it, you should just create the string of metrics directly as you loop over the metrics.

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+	return nil
+}
+
+func (t *TCPForwarder) isClosed() bool {
+	var one []byte
+	if t.conn == nil {
+		return true
+	}
+
+	t.conn.SetReadDeadline(time.Now())
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: don't you need to add some time here, otherwise the read deadline will have already passed when you try to read?

**Comment by sparrc** [comment]
> @sparrc Should we change that to the suggestion ? gathering metric before we can reach a endpoint, by not passing the first connection error ?

I don't quite understand this question, can you rephrase?

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+// Write is the default function to call to "send" a metric through the Output
+func (t *TCPForwarder) Write(metrics []telegraf.Metric) error {
+	// reconnect if needed
+	if err := t.reconnect(); err != nil {
+		return err
+	}
+	// Prepare data
+	t.Lock()
```
Comment (sparrc) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: why are you locking this?

**Comment by theist** [comment]
@tuier  Hi sorry for the noise :) the behaviour I expect does not necessarily represent the views of the telegraf dev team, I'm only a sysadm doing some research for my tick monitoring stack and a tech blog post 

I placed my findings and the config for the two telegrafs in this [gist](https://gist.github.com/theist/66fadbf0989da43577aca6196c8e37b1) 

The telegraf I'm using is a local telegraf compiled from master and with this PR merged, in a clean official golang docker container for linux 64.

All tests are made on a 4.4.0-34-generic #53-Ubuntu SMP x86_64 GNU/Linux, on localhost launching the same telegraf on foreground with different config files on different terminals

Hope it helps

**Comment by sparrc** [comment]
@tuier Can you explain the current behavior of the output plugin? What does it do if it can't connect? I generally agree with @theist that this plugin should be able to handle connection failures/restarts, whether that be in the middle of the run or at startup.

I'm a little confused by why the plugin would need to "gather" metrics, since this is an output plugin. If the connection fails then it should return an error, and telegraf's accumulator will handle buffering the points until they can successfully be written.

**Comment by tuier** [comment]
@sparrc When starting up the plugin the Connect function is call, this will return an error.
But that shouldn't stop telegraf.

I will be looking in @theist issues to see if i can find where it's coming from

**Comment by tuier** [comment]
@theist your issues should be fixed now.

@sparrc to be more precise, When telegraf start, it will call the Connect function from an output plugin. If the plugin return an error it will try a second time (after 15 second) and if it return an error again, telegraf will "fatal" and because of that stop. (https://github.com/influxdata/telegraf/blob/master/agent/agent.go#L61-L70 ,https://github.com/influxdata/telegraf/blob/master/cmd/telegraf/telegraf.go#L240-L243 )
In the Case that the TCP endpoint is not up When calling that Connect function, this plugin will return an error.

Because of this two behavior, telegraf will stop if the TCP endpoint is not set before running this plugin.

@sparrc would you prefer to ignore that error in the first Connect function ? in that case why would that function return a error, as the "fix" is to ignore that error.
Should we change the way telegraf output failure at startup and doing something like Error(print in log) but not Fatal ?

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go (hunk 9 above)
Comment (tuier) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: That the point. because of that timeout the read will error (timeout) as soon as you try to read, unless the connection is Closed in that case the error will be EOF.

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go (hunk 10 above)
Comment (tuier) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: To make sure that nothing will change the connection while trying to Write.

File: plugins/outputs/tcp_forwarder/tcp_forwarder.go (hunk 8 above)
Comment (tuier) [inline plugins/outputs/tcp_forwarder/tcp_forwarder.go]: the idea here was to leave a possibility to create/add a Join function in the serialiser interface (see line 143)

**Comment by st0ckface** [comment]
Is there any traction on this?

**Comment by tuier** [comment]
Every change requested has been made, we are just waiting for more review or a green light !

**Comment by sparrc** [comment]
@tuier can you clean up this PR? something must have gone wrong with a merge conflict, there shouldn't be 117 commits included in this PR

I would like to try to have this merged for 1.2

**Comment by sparrc** [comment]
thank you very much for this PR @tuier, but I'm closing this out in favor of https://github.com/influxdata/telegraf/pull/2094
